### ISSUE TITLE
BUG: Uninitialised Values Access

### ISSUE THREAD

**Issue by Dongcan-Jiang** [Dongcan-Jiang/matrixStats#18]
Hi @HenrikBengtsson 

I've moved the original `feature/multicore` branch to [`feature/multicore-bakup`](https://github.com/Dongcan-Jiang/matrixStats/tree/feature/multicore-backup), because it contains an  **uninitialised values access** problem.

Here is what it turned out, when I ran `make valgrind`.

```
==4049== Conditional jump or move depends on uninitialised value(s)
==4049==    at 0xC785ED9: ??? (in /usr/lib/R/library/parallel/libs/parallel.so)
==4049==    by 0x53FE33F: ??? (in /lib/x86_64-linux-gnu/libpthread-2.19.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049== 
==4049== Use of uninitialised value of size 8
==4049==    at 0xC785EDB: ??? (in /usr/lib/R/library/parallel/libs/parallel.so)https://github.com/Dongcan-Jiang/matrixStats/commits/feature/multicore
==4049==    by 0x53FE33F: ??? (in /lib/x86_64-linux-gnu/libpthread-2.19.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F433E3: ??? (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4A41F: Rf_eval (in /usr/lib/R/lib/libR.so)
==4049==    by 0x4F4BB4E: Rf_applyClosure (in /usr/lib/R/lib/libR.so)
```

However, the bug is not easy to reproduce. It rarely turns up.

Althrough I've been working on it for several days, the reason still remains unknown. Thus, I withdrew all utils implemented by myself on [`feature/multicore`](https://github.com/Dongcan-Jiang/matrixStats/commits/feature/multicore).

**Comment by HenrikBengtsson** [comment]
Odd... it looks like an issue with the 'parallel' package rather than 'matrixStats', since the error is reported in `parallel.so`.  On the other hand, I've never ran 'valgrind' on multicore code so maybe that's how memory error appear.

So, you're saying you've only observed it on your local machine and never on Travis CI, correct?  Do you have an idea if it was in a specific test script the error occurred, or when it happens it happens in a "random" test script?  If it can be reproduced by a single test script (although only once in a while), we might be able to narrow it down.

BTW, if you know a test script where it fails, you can run valgrind on that specific script as:

``` r
R --vanilla -d valgrind < tests/foo.R
```

When you say:

> Thus, I withdrew all utils implemented by myself on `feature/multicore`.

what do you mean by "all utils implemented by myself"?  Are you talking to R code or native code?   Are you saying that after withdrawing those, there is no longer a `make valgrind` problem?

**Comment by Dongcan-Jiang** [comment]
Hi @HenrikBengtsson 

> So, you're saying you've only observed it on your local machine and never on Travis CI, correct?

Yes, only on my local machine.

> Do you have an idea if it was in a specific test script the error occurred, or when it happens it happens in a "random" test script?

It only happens in either `rowRanges_parallel.R` or `rowMedian_parallel.R`. And `rowRanges_parallel.R` is more frequent.

> If it can be reproduced by a single test script (although only once in a while), we might be able to narrow it down.

However, it could not be reproduced by single test script. I've tried many ways, including removing other test files and `R --vanilla -d valgrind < rowRanges_parallel.R`, but none of them can reproduce the error.

> what do you mean by "all utils implemented by myself"? Are you talking to R code or native code? Are you saying that after withdrawing those, there is no longer a make valgrind problem?

I mean withdrawing funtions in [`utils.R`](https://github.com/Dongcan-Jiang/matrixStats/blob/feature/multicore-backup/R/utils.R), i.e. `.splitIndexRanges` and `.extractErrorMessage`.

I've run `make valgrind` on this new version for 3 times, and all finished successfully. I'm still doing more tests on it.

**Comment by Dongcan-Jiang** [comment]
Hi @HenrikBengtsson 
The problem happened in my 4th test. Thus. As you said, it may come from 'parallel' package rather than 'matrixStats'. I think it would be better to withdraw all R-level parallel codes, and only provide APIs for multiprocessing. I believe the C-level parallel  implementation will go well.

**Comment by HenrikBengtsson** [comment]
Yes, I doubt it comes from your pure R code - that's very unlikely from my experience.

I see two ways forward right now:
1. Do the R prototype implementaion. Ignore the error based on the assumption it's an issue with parallel and not us. It could also be an issue with the pthread code. It could even be a false positive by valgrind and there is no error at all. We could ask on R-devel for advice based on the valgrind output.
2. As you suggest, skip the R mockup and go straight to the C implementation that you in the past prototyped. It could still be that valgrind complaint, because the problem may be in the pthread code.  I guess we'll know eventually.

so, if you believe we should go ahead with the C implementation already now, feel free to do so. I trust your decision on this. That might also save some time.
